src/v/serde/rw/variant.h

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
```

Comment:
couple thoughts before I fully review:

- this is awesome
- I'd recommend getting a second set of eyes on it 😇

Reply:
Oh also see my note, the serde stuff is split into another PR so you can ignore it here, I'll rebase and drop the commit before merging

Reply:
ah, missed that. all good then